ctpolicy/ctpolicy.go

```diff
@@ -153,13 +177,106 @@ func (ctp *CTPolicy) GetSCTs(ctx context.Context, cert core.CertDER, expiration
 			results <- result{sct: sct}
 		}(i, g)
 	}
-	isPrecert := true
+
+	go ctp.submitPrecertInformational(cert, expiration)
+
+	var ret core.SCTDERs
+	for i := 0; i < len(ctp.groups); i++ {
+		res := <-results
+		// If any one group fails to get a SCT then we fail out immediately
+		// cancel any other in progress work as we can't continue
+		if res.err != nil {
+			// Returning triggers the defer'd context cancellation method
+			return nil, res.err
+		}
+		ret = append(ret, res.sct)
+	}
+	return ret, nil
+}
+
+// getOperatorSCTs retrieves exactly two SCTs from the total collection of
+// configured log groups, with at most one SCT coming from each group. It
+// expects that all logs run by a single operator (e.g. Google) are in the same
+// group, to guarantee that SCTs from logs in different groups do not end up
+// coming from the same operator. As such, it enforces Google's current CT
+// Policy, which requires that certs have two SCTs from logs run by different
+// operators.
+// TODO(#5938): Inline this into GetSCTs when getGoogleSCTs is removed.
+func (ctp *CTPolicy) getOperatorSCTs(ctx context.Context, cert core.CertDER, expiration time.Time) (core.SCTDERs, error) {
+	// We'll cancel this sub-context when we have the two SCTs we need, to cause
+	// any other ongoing submission attempts to quit.
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	results := make(chan result, len(ctp.operatorGroups))
+
+	// Kick off a collection of goroutines to try to submit the precert to each
+	// log operator group. Randomize the order of the groups so that we're not
+	// always trying to submit to the same two operators.
+	for i, groupIdx := range rand.Perm(len(ctp.operatorGroups)) {
+		operatorGroup := ctp.operatorGroups[groupIdx]
+		i := i
+
+		go func() {
+			// If this goroutine isn't for one of the first two groups chosen, sleep a
+			// little bit to stagger our requests to the later groups.
+			if i > 1 {
+				time.Sleep(time.Duration(i) * ctp.stagger)
+			}
+
+			// Quit if the context has errored out, most likely because two other
+			// logs from other operator groups returned SCTs already and the context
+			// has been canceled.
+			if subCtx.Err() != nil {
+				return
+			}
+
+			sct, err := ctp.race(subCtx, cert, operatorGroup, expiration)
```

Comment:
I think `race` might be overkill in the world of the new policy. In practice there is only one operator with multiple logs: Google. And that's true for the forseeable future, since (IIRC) Chrome discourages one operator running multiple logs. Rather than have a moderately-complex function that spins off goroutines at a staggered interval, maybe we should get rid of `race` in preference for a simple "timeout and try next log." The race is still happening at the higher level - between operators. Having it within an operator seems like complexity that we can afford to remove.

If we do "timeout and try next," I think it would be reasonable to set the timeout as "the remaining time left on the context, or 2 seconds, whichever is higher."